src/utils/WaitGroup.php

```diff
@@ -0,0 +1,130 @@
+<?php
+
+/*
+ *
+ *  ____            _        _   __  __ _                  __  __ ____
+ * |  _ \ ___   ___| | _____| |_|  \/  (_)_ __   ___      |  \/  |  _ \
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+ */
+
+declare(strict_types=1);
+
+namespace pocketmine\utils;
+
+use Closure;
+use InvalidArgumentException;
+use RuntimeException;
+
+/**
+ * Represents a group of tasks that can be asynchronously completed.
+ *
+ * Unlike `sync.WaitGroup` in Go, each WaitGroup can only be used once.
```

Comment:
This documentation is a bit confusing.

- Why mention golang wait group at all?
- Why can't we reuse a wait group? Seems like it shouldn't be that difficult to reset the state once the counter reaches `0`.